### ISSUE TITLE
None of these obvious methods of programmatically stopping pwsh.exe from another process result in clean{} running

### ISSUE THREAD

**Issue by alx9r** [PowerShell/PowerShell#24631]
### Steps to reproduce

The following code attempts to stop `pwsh.exe` using a variety of methods and logs which blocks are run:

```PowerShell
$(foreach ($method in {$process | Stop-Process}   ,
                      {$process.Kill()}           ,
                      {$process.CloseMainWindow()},
                      {taskkill /PID $process.Id } ) {
    {
        begin {
            $logFile = "$PID`_log.txt"
            'begin' | Set-Content $logFile
        }
        end {
            foreach ($i in 1..10) {
                Start-Sleep -Milliseconds 100
            }
            'end' | Add-Content $logFile
        }
        clean {
            'clean' | Add-Content $logFile
        }
    } |
        Set-Content .\probe.ps1

    $process = Start-Process -FilePath pwsh.exe -ArgumentList .\probe.ps1 -PassThru
    $logFile = "$($process.Id)`_log.txt"
    do { Start-Sleep -Milliseconds 10 }
    while (-not (Test-Path $logFile))

    $output = . $method

    do { Start-Sleep -Milliseconds 10 }
    while (Get-Process -Id $process.Id -ErrorAction SilentlyContinue)

    [pscustomobject]@{
        Id       = $process.Id
        method   = $method
        begin    = ($begin = ($content = Get-Content $logFile) -contains 'begin')
        end      = ($end = $content -contains 'end')
        clean    = ($clean = $content -contains 'clean')
        signaled = $begin -and (-not $end)
        result   = if ($signaled -and $clean) {'success'} else {'failure'}
        output   = $output
    }
}) |
    Format-Table
```

The actual behavior seems to indicate that `pwsh.exe` either
- did not receive a stop signal, or
- received a stop signal but terminated without invoking `clean{}`.

### Expected behavior

```console
   Id method                     begin   end clean signaled result  output
   -- ------                     -----   --- ----- -------- ------  ------
24888 $process | Stop-Process     True False False     True failure
16076 $process.Kill()             True False False     True failure
 2164 $process.CloseMainWindow()  True False  True     True success
16848 taskkill /PID $process.Id   True False  True     True success
```

### Actual behavior

```console
   Id method                     begin   end clean signaled result  output
   -- ------                     -----   --- ----- -------- ------  ------
24888 $process | Stop-Process     True False False     True failure
16076 $process.Kill()             True False False     True failure
 2164 $process.CloseMainWindow()  True  True  True    False failure False
16848 taskkill /PID $process.Id   True  True  True    False failure SUCCESS: Sent termination signal to the process with PID 16848.
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.6
PSEdition                      Core
GitCommitId                    7.4.6
OS                             Microsoft Windows 10.0.22631
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by rhubarb-geek-nz** [comment]
I think it is important to understand what mechanism is being used at the operating system level.

The only clean way for a process to terminate is if itself calls [ExitProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-exitprocess) as a result of its own execution flow. [Terminating a Process](https://learn.microsoft.com/en-us/windows/win32/procthread/terminating-a-process), this could be as a result of returning from WinMain/main handling a windows message or a console event, etc.

If the mechanism involves [TerminateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess) then this not a clean exit and the process itself, including shared libraries, has no opportunity to clean itself up.

**Comment by jborean93** [comment]
This is a difficult topic but TLDR I don't think PowerShell can do anything here as it is Windows and other environment specific things that control how these signals are processed.

Using `Stop-Process` or `Process.Kill()` calls [TerminateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess) which is a hard exit of a process which Windows does not provide a public way to hook into so PowerShell cannot do any cleanup work.

The `taskkill.exe` (without `/F`) and `Process.CloseMainWindow()` work by sending the [WM_CLOSE](https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-close) message to the main window associated with the process. For GUI based applications this is easy but for console applications like PowerShell there may or may not be a window handle associated with it.

When you spawn PowerShell in Windows Terminal you'll find that there is no window handle as it is using a Console PTY and PowerShell itself doesn't actually have a conhost handle associated with it. You can see that `(Get-Process -Id $pid).MainWindowHandle` is 0 (null/no handle) when PowerShell is either spawned where the default console application is Windows Terminal (top) or spawned with `wt.exe pwsh.exe` (bottom)

![image](https://github.com/user-attachments/assets/46ebecde-6c98-40d4-8fb5-9b3f9bc25a15)

If there is no window handle then PowerShell will never received anything so just continues to run like normal. You could instead find the Windows Terminal process and call `CloseMainWindow()` on that and I'm pretty sure that will close Windows Terminal and PowerShell but that's not ideal if you are targeting just PowerShell.

When Windows is configured to run PowerShell through the old conhost console application or was spawned with conhost you can see that there is now a window handle associated with the process.

![image](https://github.com/user-attachments/assets/a6f0e24c-6c3f-4c24-a564-80c50c15bf33)

You can see that when using the old conhost console that `taskkill` and `Process.CloseMainWindow()` now works as expected and even the `clean` block is processed. This is because PowerShell is now able to see the close event and it will stop the current pipeline.

```powershell
$sbk = {
    [CmdletBinding()]
    param()

    end {
        Set-Content test.log "Running"
        Start-Sleep -Seconds 60
        Add-Content test.log "End"
    }
    clean {
        Add-Content test.log "Clean"
    }
}
Set-Content test.ps1 $sbk

$proc = Start-Process pwsh.exe -ArgumentList '-File test.ps1' -PassThru
Start-Sleep -Seconds 2
$proc.CloseMainWindow()
Get-Content test.log
```

![image](https://github.com/user-attachments/assets/06201309-d838-447d-963e-38f337aa06a0)

Now to achieve what you are looking for you need to in fact find a way to send the CTRL+C signal to the console which in turn is able to pass through that signal to PowerShell to process. This works for both the old conhost and new ConPTY (Windows Terminal). Unfortunately Windows does not have a builtin way to signal a target process yet there is technically a way.

You can call [AttachConsole](https://learn.microsoft.com/en-us/windows/console/attachconsole) to attach the console of a specific process to the current process and call [GenerateConsoleCtrlEvent](https://learn.microsoft.com/en-us/windows/console/generateconsolectrlevent) on the current console which also means it will be sent to the PowerShell process that is attached to the same console. Because you need to free the existing console and that `AttachConsole` are global actions in the process you cannot safely do this in PowerShell itself because it can impact other things running in the process or future events if the original console handle was freed during the operation. Luckily you can create a custom executable that does just this through [kill.ps1](https://gist.github.com/jborean93/c04426e88d62d09f539088f5bd54bc7d).

```powershell
$sbk = {
    [CmdletBinding()]
    param()

    end {
        Set-Content test.log "Running"
        Start-Sleep -Seconds 60
        Add-Content test.log "End"
    }
    clean {
        Add-Content test.log "Clean"
    }
}
Set-Content test.ps1 $sbk

$proc = Start-Process pwsh.exe -ArgumentList '-File test.ps1' -PassThru
Start-Sleep -Seconds 2
.\kill.exe $proc.Id
Get-Content test.log
```

![image](https://github.com/user-attachments/assets/4c562a63-d6c0-45ae-af49-8849791bd80d)

**Comment by alx9r** [comment]
@jborean93 

Thank you very much for your helpful reply.  My summary interpretation of the situation is as follows:

1. There is no reliable windows-wide standard way to request the PowerShell process to gracefully stop.
2. I should design my programs to assume they might be terminated without a chance to clean up.

This all has me wondering whether there ought to be an in-built way for one PowerShell peer to request another PowerShell peer to stop gracefully.  Such graceful stopping currently occurs for `Stop-Job`.  It's not obvious to me why graceful stopping shouldn't also happen for a command like `Stop-Process pwsh`.

**Comment by rhubarb-geek-nz** [comment]
> Such graceful stopping currently occurs for `Stop-Job`. It's not obvious to me why graceful stopping shouldn't also happen for a command like `Stop-Process pwsh`.

When you use Start-Process the program that is run could be absolutely anything, so the commands that work with the noun Process only deals with generic windows processes. When PowerShell manages Jobs it already knows that it is dealing with jobs that it created and can deal with them accordingly.

An alternative is to use jobs when you really want life and death control over the execution.

**Comment by jborean93** [comment]
Yep with a job or anything PSRemoting based, PowerShell can assume the target is PowerShell and has the ability to send the [PowerShell.Stop](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.powershell.stop?view=powershellsdk-7.4.0#system-management-automation-powershell-stop) signal to stop the pipeline. Even then it's not guaranteed that this will complete because PowerShell may be busy running something so it can't respond to the signal but it's the only graceful way to do so.

This is unlike stopping a process where the only means available to you are:

+ `taskkill.exe /PID ...` or the `Process.CloseMainWindow()` and hope there is a main window to respond and that it does work
+ use the `kill.exe` (generated by [kill.ps1](https://gist.github.com/jborean93/c04426e88d62d09f539088f5bd54bc7d) to send CTRL+C to the console and hope that it works
+ use `Stop-Process` to terminate it completely

Depending on the exe and what it is running some of these examples may or may not work but there is no generic catch all that will always work outside of `Stop-Process` to do a hard kill really.

**Comment by alx9r** [comment]
>...the commands that work with the noun Process only deals with generic windows processes...

Hmm...I don't think this is accurate.  `Stop-Process` seems to special some processes.  Here is the special-case for `SVCHOST`:

https://github.com/PowerShell/PowerShell/blob/c24e8e440242d16d46394ba0e7778258f39fa6f4/src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs#L1225-L1229

If that's prudent, then why wouldn't a special-case for a `pwsh` process be prudent?

**Comment by jborean93** [comment]
> If that's prudent, then why wouldn't a special-case for a pwsh process be prudent?

How can it, there is no easy API in Windows to send ctrl+c to a process and the way to do it requires a completely separate executable as the API calls needed to do so mutate process wide things and could have negative side effects after running (lost the original console). Even then ctrl+c isn't guaranteed to have stopped the process as it could be busy running something to respond to the stop. The `SVCHOST` example is somewhat special in that stopping a service will actually stop the process in the end.

**Comment by alx9r** [comment]
>How can it, there is no easy API in Windows to send ctrl+c...

In principle, at least, `pwsh.exe` could listen for a stop request signal on any of the various IPC channels and interpret and act on that in the same way as it does ctrl+c.

**Comment by jborean93** [comment]
But it doesn't right now and even if such an interface exists there's no guarantee it will respond to it. Try pressing ctrl+c after running `[System.Threading.Thread]::Sleep(10000)` and you'll see it won't respond until after the function returns.

If you want a way to cancel a PowerShell process and are happy to be subject to the limitations of the stop signal then use jobs with `Stop-Job` or the [NamedPipeConnectionInfo](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.runspaces.namedpipeconnectioninfo?view=powershellsdk-7.4.0) runspace connection info as a `Runspace` target in which you can also send the stop signal.

**Comment by alx9r** [comment]
>If you want a way to cancel a PowerShell process then use...the NamedPipeConnectionInfo runspace connection info as a Runspace target in which you can also send the stop signal...

I _think_ you are suggesting that I should be able to connect to an unrelated running instance of `pwsh` by way of `NamedPipeConnectionInfo` and somehow send a stop signal to all of the running runspaces in that process.  Am I understand that correctly?  If so, is there an example you could point me to that does something like that?

**Comment by jborean93** [comment]
You can stop the immediate runspace on that connection but doing so for other running runspaces is not exposed through any public API and would require reflection to pull off. Maybe something like this could work when running in the target process but I've not tested it and once again you cannot guarantee that it will not block. It also closes the runspace fully vs just stopping the currently running pipeline of that runspace.

```powershell
Get-Runspace | ForEach-Object {
    if ($_.Id -eq [Runspace]::DefaultRunspace.Id) {
        return
    }

    $_.Close()
}
```

Ultimately this leads to point where if you want to stop a PowerShell process the only stable way available to you is the hard killing `Stop-Process`. You can try and various workarounds to potentially work around it but it is never guaranteed to actually stop the pwsh process.

**Comment by jborean93** [comment]
If you are wanting to ensure something is cleaned the only main thing that outlives a process is probably files you manage. Other things like handles to various Win32 objects are automatically closed by Windows when it ends the process so regardless of whether you are explicitly disposing it in PowerShell's clean block or not it will be cleaned up even if hard killing the process.

Files are tricky, if dealing with temporary files you either just need to accept the fact that the process can leave temp files laying around and to use the temp path so that they are cleaned up eventually. In most cases your process will end normally and they will be removed. If you have a scenario where you are routinely calling `Stop-Process` on a target process then you should look at using some process wide synchronisation to ensure that the code can respond to your custom signal or you use files in a way that they themselves are cleaned up after all their HANDLES are closed. For example using a [FileStream](https://learn.microsoft.com/en-us/dotnet/api/system.io.filestream?view=net-9.0) opened with [FileStreamOptions.Options.DeleteOnClose](https://learn.microsoft.com/en-us/dotnet/api/system.io.fileoptions?view=net-9.0) will ensure Windows deletes the file once all HANDLEs to that file are closed, including when the process is hard killed.

**Comment by alx9r** [comment]
>...it is never guaranteed to actually stop the pwsh process

I didn't intend to make the impression that I'm looking for a guarantee for all scenarios.  Rather, I'm looking for a method to gracefully stop an independent running instance of a `pwsh` process running my code.  Obviously, my code running in that instance will have to cooperate to allow that to happen.  The point of my OP is that the obvious methods of stopping `pwsh` don't do that.

**Comment by jborean93** [comment]
> Rather, I'm looking for a method to gracefully stop an independent running instance of a pwsh process running my code

That's my point, you can send the signal for it to "gracefully stop" through things like my `kill.exe` example but there are no guarantees that it will actually do so.

What you want just doesn't exist at the process level. There are workarounds that may or may not work depending on your scenario but anything you come up with is going to be tailored to what you are doing. Technically PowerShell could try and add something to respond to external requests to stops but

+ it will be Windows specific
+ anything implemented may not even work due to the pipeline being busy
+ now has to consider access controls and who can actually send those signals
+ needs someone to actually design and implement it

If you need such a thing then you'll have to implement it in your script through IPC primitives like named mutexes, events, semaphores and so on. Your script can spawn nested pipelines or thread jobs and wait for those IPC primitives in the main thread so that once a signal is sent you can cancel what you need accordingly.

**Comment by alx9r** [comment]
FWIW, I managed a working demo [here](https://stackoverflow.com/a/79248980/1404637).  In principle `pwsh` could listen out-of-the-box for a stop request event like that demo.  The technique used in the demo listens for an event using `EventWaitHandle()` which is apparently ["supported on all platforms"](https://apisof.net/catalog/07a9b8290110ba12610f783e5e8d625d).  The demo uses `GenerateConsoleCtrlEvent` which is a Windows-only API.  Surely, though, a handler with internal access to PowerShell could patch that event to do whatever `SIGINT` or `CTRL_C_EVENT` does on each platform.  

In other words, reliably signaling a `pwsh` process to try to perform a graceful shutdown before forcibly terminating that process seems to be technically possible in a general way.  Perhaps there are reasons not to build that capability in to `pwsh`.  It seems to me, though, that if allowing for graceful stopping is prudent for `Stop-Job`, then it is also prudent for stopping any `pwsh` process.  In both cases there is always the option to escalate to forcible termination if graceful shutdown doesn't complete in a timely manner.

**Comment by jborean93** [comment]
> The demo uses GenerateConsoleCtrlEvent which is a Windows-only API. Surely, though, a handler with internal access to PowerShell could patch that event to do whatever SIGINT or CTRL_C_EVENT does on each platform.

That's what my `kill.exe` example does, it loads the console of the target process id and sends `CTRL_C_EVENT` so anything else sharing that console receives that signal. It uses a sub-process because you cannot safely detach your current console, attach the target one, then re-attach the original console and those actions are process wide so can affect other things that might be running. You can use the `NamedPipeConnectionInfo` to attach to a target PowerShell process and just run it inside that process without having to mange the console changes:

```powershell
Function Stop-PowerShellProcess {
    [CmdletBinding()]
    param(
        [Parameter(Mandatory)]
        [int]
        $ProcessId
    )

    $connInfo = [System.Management.Automation.Runspaces.NamedPipeConnectionInfo]::new($ProcessId)
    $runspace = [RunspaceFactory]::CreateRunspace($connInfo)
    try {
        $runspace.Open()
        $ps = [PowerShell]::Create($runspace)
        $ps.AddScript(@'
Add-Type -Namespace Kernel32 -Name Native -MemberDefinition @"
[DllImport("kernel32.dll", EntryPoint = "GenerateConsoleCtrlEvent")]
private static extern bool GenerateConsoleCtrlEventNative(
    int dwCtrlEvent,
    int dwProcessGroupId);

public static void GenerateConsoleCtrlEvent(int eventId, int processGroupId)
{
    if (!GenerateConsoleCtrlEventNative(eventId, processGroupId))
    {
        throw new System.ComponentModel.Win32Exception();
    }
}
"@

[Kernel32.Native]::GenerateConsoleCtrlEvent(0, 0)
'@).Invoke()
        foreach ($e in $ps.Streams.Error) {
            Write-Error -ErrorRecord $e
        }
    }
    finally {
        $runspace.Dispose()
    }
}
```

**Comment by kasini3000** [comment]
Hi boy, I'm here to discuss.
I don't like this kind of code:
Register-EngineEvent -SourceIdentifier Powershell.Exiting -Action { Write-Host "bye" }

I suggest:
1.1 Add a parameter to pwsh -SIGKILL {}, which is empty by default
1.2 Add a callback method to pwsh : $pid.SIGKILL{}, is empty by default.
The two above point to the same function. Modify any one above, and the other will also change.

1.3 Add a event to pwsh : $pid.SIGKILL()

1.4 Modify Clean Logic:
When the SIGKILL {} method is not empty, and  $pid.SIGKILL() event triggered from another process,
Each thread stops and executes clean-block logic, then executes $pid.SIGKILL() 

send SIGKILL directly to PWSH, then there is no need to consider Ctrl+C .

**Comment by rhubarb-geek-nz** [comment]
> Each thread stops and executes clean-block logic

This would not work as each thread would need to be cancelled within the thread to unwind the nested try/finally statements within the current execution stack. This is then is a problem because you could have blocking calls within a finally block.

The pthreads approach to this is that you can't call blocking API within a cancellation handler but there is no such restriction with .NET threads

Although the threading execution model is pre-emptive, the thread unwinding model is still cooperative. You can't make something cooperate, it has to do it voluntarily. Using TerminateThread is arguably worse than calling TerminateProcess because you risk corrupting the process heap and contained objects.

**Comment by alx9r** [comment]
>This probably works normally but could have the chance of stopping the runspace exchange due to CTRL+C cancelling the runspace before it returns back the results but I never encountered that during testing.

@jborean93 Using [this incarnation of `Stop-PowerShellProcess`](https://github.com/PowerShell/PowerShell/issues/24631#issuecomment-2515472936) with a process id of a running job seems to cause the calling session to irrecoverably lock up.

**Comment by jborean93** [comment]
Yep so caveats apply, use the kill example to avoid all that.

**Comment by jborean93** [comment]
Yep so caveats apply, use the kill executable example to avoid all that.

**Comment by kasini3000** [comment]
@rhubarb-geek-nz 
Elegantly stopping pwsh processes and attached threads is a necessary feature of a language.
We must find ways to  achieve it, It also needs to be easy to use.
Otherwise, 
PowerShell lags too much behind Python and Golang. We don't want to see this.

**Comment by rhubarb-geek-nz** [comment]
> ... lags too much behind Python and Golang.

Same rules apply to Python and golang. All that can be done is signal a thread/goroutine to terminate and let it throw an exception. If it is unresponsive nothing will happen, it is cooperative.

**Comment by alx9r** [comment]
Santiago Squarzon provided [this technique](https://stackoverflow.com/a/79249620/1404637) that connects to a running PowerShell process and operates directly on runspaces.  After quite a bit of testing I have found that was the only technique that had all of the following merits:

- works on both job and non-job runspaces
- doesn't require any special listening configured on the target
- works across all combinations of Windows PowerShell and PowerShell 7
